File: Question 1219 - Path with Maximum Gold/src/Solution.java
class Solution
{
    private int solve(int[][] grid, boolean[][] visited, int r, int c, int[][] dir, int gold)
    {
        visited[r][c] = true;
        int maxGold = gold;

        for(int[] d : dir)
        {
            int newX = r + d[0];
            int newY = c + d[1];

            if(newX >= 0 && newX < grid.length && newY >= 0 && newY < grid[0].length && !visited[newX][newY] && grid[newX][newY] != 0)
            {
                maxGold = Math.max(maxGold, grid[newX][newY] + solve(grid, visited, newX, newY, dir, gold));
            }
        }

        visited[r][c] = false;
        return maxGold;
    }
    public int getMaximumGold(int[][] grid)
    {
        int r = grid.length;
        int c = grid[0].length;
        boolean[][] visited = new boolean[r][c];

        int[][] dir = {{-1,0},{1,0},{0,1},{0,-1}};
        int currGold = 0;
        int maxGold = Integer.MIN_VALUE;

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if(grid[i][j] != 0)
                {
                    currGold = solve(grid,visited,i,j,dir,grid[i][j]);
                }
                maxGold = Math.max(currGold,maxGold);
            }
        }
        return maxGold;
    }
}

File: Question 1219 - Path with Maximum Gold/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
    int solve(vector<vector<int>> &grid, vector<vector<bool>> &visited, int i, int j, int gold, vector<vector<int>> &dir)
    {
        visited[i][j] = true;
        int maxGold = gold;

        for (auto &d : dir)
        {
            int x = i + d[0];
            int y = j + d[1];

            if (x >= 0 && x < grid.size() && y >= 0 && y < grid[0].size() && !visited[x][y] && grid[x][y] != 0)
            {
                maxGold = max(maxGold, grid[x][y] + solve(grid, visited, x, y, gold, dir));
            }
        }

        visited[i][j] = false;
        return maxGold;
    }

public:
    int getMaximumGold(vector<vector<int>> &grid)
    {
        int r = grid.size(), c = grid[0].size();

        int currGold = 0, maxGold = INT_MIN;
        vector<vector<bool>> visited(r, vector<bool>(c, false));
        vector<vector<int>> dir = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (int i = 0; i < r; i++)
        {
            for (int j = 0; j < c; j++)
            {
                if (!visited[i][j])
                {
                    currGold = solve(grid, visited, i, j, grid[i][j], dir);
                }

                maxGold = max(maxGold, currGold);
            }
        }

        return maxGold;
    }
};